src/imputors.jl

```diff
@@ -1,77 +1,132 @@
 """
     Imputor
 
-An imputor stores information about imputing values in `AbstractArray`s and `DataFrame`s.
+An imputor stores information about imputing values in `AbstractArray`s and `Tables.table`s.
 New imputation methods are expected to sutype `Imputor` and, at minimum,
-implement the `impute!{T<:Any}(imp::<MyImputor>, ctx::Context, data::AbstractArray{T, 1})`
-method.
+implement the `impute!(imp::<MyImputor>, data::AbstractVector)` method.
 """
-
 abstract type Imputor end
 
-"""
-    impute!(imp::Imputor, data::Dataset, limit::Float64=0.1)
+# A couple utility methods to avoid messing up var and obs dimensions
+obsdim(imp::Imputor) = imp.vardim == 1 ? 2 : 1
+vardim(imp::Imputor) = imp.vardim
 
-Creates a `Context` using information about `data`. These include
+function obswise(imp::Imputor, data::AbstractMatrix)
+    (imp.vardim == 1 ? view(data, :, i) : view(data, i, :) for i in axes(data, obsdim(imp)))
+end
 
-1. missing data function which defaults to `missing`
+function varwise(imp::Imputor, data::AbstractMatrix)
+    (imp.vardim == 1 ? view(data, i, :) : view(data, :, i) for i in axes(data, vardim(imp)))
+end
 
-2. number of elements: `*(size(data)...)`
+function filterobs(f::Function, imp::Imputor, data::AbstractMatrix)
+    mask = [f(x) for x in obswise(imp, data)]
+    return imp.vardim == 1 ? data[:, mask] : data[mask, :]
+end
 
-# Arguments
-* `imp::Imputor`: the Imputor method to use
-* `data::Dataset`: the data to impute
-* `limit::Float64: missing data ratio limit/threshold (default: 0.1)`
+function filtervars(f::Function, imp::Imputor, data::AbstractMatrix)
+    mask = [f(x) for x in varwise(imp, data)]
+    return imp.vardim == 1 ? data[mask, :] : data[:, mask]
+end
+
+"""
+    impute(data, imp::Imputor)
 
-# Return
-* `Dataset`: the input `data` with values imputed.
+Returns a new copy of the `data` with the missing data imputed by the imputor `imp`.
 """
-function impute!(imp::Imputor, data::Dataset, limit::Float64=0.1)
-    ctx = Context(*(size(data)...), 0, limit, ismissing)
-    return impute!(imp, ctx, data)
+function impute(data, imp::Imputor)
+    # Call `deepcopy` because we can trust that it's available for all types.
+    return impute!(deepcopy(data), imp)
 end
 
+
+# This is a necessary fallback because the tables method doesn't have a type declaration
+impute!(data::AbstractVector, imp::Imputor) = MethodError(impute!, (data, imp))
```

Comment:
I think it would be weird to throw a `MethodError` from `impute!(data, imp)` and I don't think an `ArgumentError` for this situation would be more confusing.

Reply:
It's actively documented that we expect people to implement a vector based imputation method.

Reply:
Most important point here is not about error types.
It is about workin `Tables.RowTable`
Which this will current error on AFAICT.
And for no good reason since it gets converted to a column table inside the main function anyway

Reply:
Would you be okay with throwing a `MethodError` from the table method?